File: moderate/jolly_jumpers.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func main() {
	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()
	scanner := bufio.NewScanner(data)
	scanner.Split(bufio.ScanWords)
	for scanner.Scan() {
		var (
			n, v, vp int
			u        []bool
		)
		jolly, done := true, false
		fmt.Sscan(scanner.Text(), &n)
		if n == 1 {
			done = true
		} else {
			u = make([]bool, n-1)
		}
		scanner.Scan()
		fmt.Sscan(scanner.Text(), &vp)
		for i := 1; i < n; i++ {
			scanner.Scan()
			fmt.Sscan(scanner.Text(), &v)
			if done {
				continue
			}
			x := abs(v - vp)
			if x >= n || x == 0 || u[x-1] {
				jolly, done = false, true
			} else {
				u[x-1], vp = true, v
			}
		}
		if jolly {
			fmt.Println("Jolly")
		} else {
			fmt.Println("Not jolly")
		}
	}
}

File: moderate/jolly_jumpers.rb
File.open(ARGV[0]).each_line do |line|
  s = line.chomp.split.map(&:to_i)
  jolly, n = true, s[0]
  if n > 1 then
    u = Array.new(n-1, false)
    (2..n).each do |i|
      x = (s[i] - s[i-1]).abs
      if x >= n or x == 0 or u[x-1] then
        jolly = false
        break
      else
        u[x-1] = true
      end
    end
  end
  puts jolly ? "Jolly" : "Not jolly"
end
